precompile/assetsbridge/bridge_out.go

```diff
@@ -0,0 +1,516 @@
+	authzKeeper  AuthzKeeper
+}
+
+func newBridgeOutMethod(
+	bridgeKeeper BridgeKeeper,
+	evmKeeper EvmKeeper,
+	authzKeeper AuthzKeeper,
+) *BridgeOutMethod {
+	return &BridgeOutMethod{
+		bridgeKeeper: bridgeKeeper,
+		evmKeeper:    evmKeeper,
+		authzKeeper:  authzKeeper,
+	}
+}
+
+func (m *BridgeOutMethod) MethodName() string {
+	return BridgeOutMethodName
+}
+
+func (m *BridgeOutMethod) MethodType() precompile.MethodType {
+	return precompile.Write
+}
+
+func (m *BridgeOutMethod) RequiredGas(_ []byte) (uint64, bool) {
+	// Fallback to the default gas calculation.
+	return 0, false
+}
+
+func (m *BridgeOutMethod) Payable() bool {
+	return false
+}
+
+func (m *BridgeOutMethod) Run(
+	context *precompile.RunContext,
+	rawInputs precompile.MethodInputs,
+) (precompile.MethodOutputs, error) {
+	// extract inputs
+	inputs, err := m.extractInputs(rawInputs)
+	if err != nil {
+		return precompile.MethodOutputs{false}, err
+	}
+
+	// run validation
+	if err := m.validate(context, inputs); err != nil {
+		return precompile.MethodOutputs{false}, err
+	}
+
+	return m.execute(context, inputs)
+}
+
+// execute will execute the burn of the token then send
+// AssetsUnlocked events to the bridgeKeeper.
+func (m *BridgeOutMethod) execute(
+	context *precompile.RunContext,
+	inputs *bridgeOutInputs,
+) (precompile.MethodOutputs, error) {
+	var (
+		err            error
+		assetsUnlocked *bridgetypes.AssetsUnlockedEvent
+		isBTC          = bytes.Equal(
+			common.HexToAddress(evmtypes.BTCTokenPrecompileAddress).Bytes(),
+			inputs.Token.Bytes(),
+		)
+	)
+
+	switch inputs.Chain {
+	case TargetChainEthereum:
+		if isBTC {
+			assetsUnlocked, err = m.executeBitcoin(context, inputs)
+		} else {
+			assetsUnlocked, err = m.executeEthereum(context, inputs)
+		}
+	case TargetChainBitcoin:
+		assetsUnlocked, err = m.executeBitcoin(context, inputs)
+	}
+
+	if assetsUnlocked != nil {
+		err := context.EventEmitter().Emit(
+			NewAssetsUnlockedEvent(
+				assetsUnlocked.UnlockSequence.BigInt(),
+				assetsUnlocked.Recipient,
+				common.HexToAddress(assetsUnlocked.Token),
+				context.MsgSender(),
+				uint8(assetsUnlocked.Chain), //nolint:gosec // G115: Safe conversion, Chain is validated elsewhere
+				assetsUnlocked.Amount.BigInt(),
+			),
+		)
+		if err != nil {
+			return nil, fmt.Errorf("failed to emit AssetsUnlocked event: [%w]", err)
+		}
+
+	}
+
+	return precompile.MethodOutputs{err == nil}, err
+}
+
+func (m *BridgeOutMethod) executeEthereum(
+	context *precompile.RunContext,
+	inputs *bridgeOutInputs,
+) (*bridgetypes.AssetsUnlockedEvent, error) {
+	var (
+		sdkCtx          = context.SdkCtx()
+		bridgeAddrBytes = common.HexToAddress(
+			evmtypes.AssetsBridgePrecompileAddress,
+		).Bytes()
+		spenderAddr = context.MsgSender()
+	)
+
+	sdkAmount, err := precompile.TypesConverter.BigInt.ToSDK(inputs.Amount)
+	if err != nil {
+		return nil, fmt.Errorf("unable to convert amount to sdk type: %v", err)
+	}
+
+	call, err := evmtypes.NewERC20BurnFromCall(
```

Comment:
Thoughts on moving the burn call to `x/bridge` and expose it as `BurnERC20` from the keeper? We do that for Bitcoin. Also, minting happens entirely within `x/bridge`. We would have a consistent approach were the precompile is just an interface and the `x/bridge` module is responsible for the logic.